target/linux/ath79/dts/ar7100_mikrotik_routerboard-4xx.dtsi

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: GPL-1.0-or-later OR MIT
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "ar7100.dtsi"
+
+/ {
+	aliases {
+		led-boot = &led_user;
+		led-failsafe = &led_user;
+		led-running = &led_user;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_user: user {
+			label = "mikrotik:yellow:user";
+			gpios = <&gpio 4 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+
+		led1 {
+			label = "mikrotik:green:led1";
+			gpios = <&cpld_gpio 0 GPIO_ACTIVE_LOW>;
+		};
+
+		led2 {
+			label = "mikrotik:green:led2";
+			gpios = <&cpld_gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+		led3 {
+			label = "mikrotik:green:led3";
+			gpios = <&cpld_gpio 1 GPIO_ACTIVE_LOW>;
+		};
+
+		led4 {
+			label = "mikrotik:green:led4";
+			gpios = <&cpld_gpio 3 GPIO_ACTIVE_LOW>;
+		};
+
+		led5 {
+			label = "mikrotik:green:led5";
+			gpios = <&cpld_gpio 8 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&gpio {
+	ngpios = <31>;
+	gpio-line-names =
+		"",     "", "", "", "LED", "RDY",  "",  "MDC",
+		"MDIO", "", "", "", "",    "",     "",  "",
+		"",     "", "", "", "",    "",     "",  "",
+		"",     "", "", "", "",    "",     "",  "";
+};
+
+&pcie0 {
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+
+	compatible = "mikrotik,rb4xx-spi";
+
+	flash@0 {
+		compatible = "pm25lv512", "jedec,spi-nor";
+
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "mikrotik,routerboot-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "routerboot1";
+				reg = <0x0 0x0>;
+				read-only;
+			};
+
+			hard_config {
+				label = "hard_config";
+				read-only;
+			};
+
+			bios {
+				label = "bios";
+				size = <0x1000>;
+				read-only;
+			};
+
+			routerboot2 {
+				label = "routerboot2";
+				read-only;
+			};
+
+			soft_config {
+				label = "soft_config";
+			};
+		};
+	};
+
+	cpld@1 {
+		compatible = "mikrotik,rb4xx-cpld";
+
+		reg = <1>;
+		spi-max-frequency = <25000000>;
+
+		cpld_gpio: gpio {
+			compatible = "mikrotik,rb4xx-gpio";
+
+			base = <32>;
+
+			gpio-controller;
+			#gpio-cells = <2>;
+			gpio-line-names =
+				"LED1", "LED2", "LED3", "LED4",
+				"FAN",  "ALE",  "CLE",  "nCE",
+				"LED5";
+		};
+
+		nand {
+			compatible = "mikrotik,rb4xx-nand";
+
+			gpios = <&cpld_gpio 5 GPIO_ACTIVE_HIGH>, // ALE
+				<&cpld_gpio 6 GPIO_ACTIVE_HIGH>, // CLE
+				<&cpld_gpio 7 GPIO_ACTIVE_HIGH>, // nCE
+				<&gpio      5 GPIO_ACTIVE_HIGH>; // RDY
+
+			partitions {
+				compatible = "fixed-partitions";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				partition@0 {
+					label = "booter";
+					reg = <0x0000000 0x0040000>;
+					read-only;
+				};
+
+				partition@40000 {
+					label = "kernel";
+					reg = <0x0040000 0x03c0000>;
+				};
+
+				partition@400000 {
+					label = "ubi";
+					reg = <0x0400000 0x7c00000>;
+				};
```

Comment:
I'm wondering why we hardcode the kernel and root partitions sizes in dts, seems like a bad idea?

Just looking at the target makefile, it seems that the image is built similarly as the NOR ones, i.e. the kernel and rootfs are embedded in yaffs (via kernel2minor), so I *think* it should be possible to use the minor splitter? (this will probably require adjusting the sysupgrade scripts but it seems to be a good time to do this while we bootstrap mikrotik to ath79: @adrianschmutzler @xback opinons?)

Reply:
OK forget this for now, this can't be implemented in the current nand upgrade framework. I'll discuss with @rmilecki later, in the mean time this comment can be ignored.